File: src/player/queue.rs
//! Framework-free queue model.
//!
//! Port of `web/player-core/src/queue.ts` (BSD-3-Clause). Canonical items +
//! cursor, with repeat modes and shuffle layered over the canonical list
//! (never destroying it):
//!
//! - repeat `one` is resolved by the player at EOS (reload current), so
//!   `next` treats it like `off`;
//! - repeat `all` wraps `next` to index 0;
//! - shuffle produces a presentation order via [`super::order`] (seeded RNG);
//!   [`QueueModel::previous`] consults a history stack so Previous works
//!   under shuffle; toggling shuffle off restores canonical order.

use super::order::{Rng, next_index_under_repeat, shuffle_order};
use super::types::{PlaybackItem, RepeatMode};

/// Observable queue state.
#[derive(Debug, Clone, PartialEq)]
pub struct QueueState {
    /// Canonical item list.
    pub items: Vec<PlaybackItem>,
    /// Cursor index (`-1` when empty).
    pub index: i64,
}

/// Framework-free queue model.
pub struct QueueModel {
    items: Vec<PlaybackItem>,
    index: i64,
    order: Vec<usize>,
    history: Vec<i64>,
    rng: Box<Rng>,
    repeat: RepeatMode,
    shuffling: bool,
}

impl QueueModel {
    /// Creates an empty queue with the given shuffle RNG.
    pub fn new(rng: Box<Rng>) -> Self {
        Self {
            items: Vec::new(),
            index: -1,
            order: Vec::new(),
            history: Vec::new(),
            rng,
            repeat: RepeatMode::Off,
            shuffling: false,
        }
    }

    /// Snapshot of the observable state.
    pub fn state(&self) -> QueueState {
        QueueState {
            items: self.items.clone(),
            index: self.index,
        }
    }

    /// The canonical item list.
    pub fn items(&self) -> &[PlaybackItem] {
        &self.items
    }

    /// The cursor index (`-1` when empty).
    pub fn index(&self) -> i64 {
        self.index
    }

    /// The active repeat mode.
    pub fn repeat(&self) -> RepeatMode {
        self.repeat
    }

    /// Whether shuffle is on.
    pub fn shuffling(&self) -> bool {
        self.shuffling
    }

    /// Item at `i`, or `None`.
    pub fn at(&self, i: i64) -> Option<&PlaybackItem> {
        if i < 0 {
            return None;
        }
        self.items.get(i as usize)
    }

    /// The current item, or `None`.
    pub fn current(&self) -> Option<&PlaybackItem> {
        self.at(self.index)
    }

    /// Replaces the canonical list and cursor (session restore).
    pub fn set_all(&mut self, items: Vec<PlaybackItem>, index: i64) {
        self.items = items;
        self.index = index;
    }

    /// Sets the repeat mode.
    pub fn set_repeat(&mut self, mode: RepeatMode) {
        self.repeat = mode;
    }

    /// Toggles shuffle; rebuilds the current-first presentation order.
    pub fn set_shuffle(&mut self, on: bool) {
        if self.shuffling == on {
            return;
        }
        self.shuffling = on;
        self.history.clear();
        self.rebuild_order();
    }

    /// Test/inspection seam: the active presentation order (indices).
    pub fn presentation_order(&self) -> Vec<usize> {
        if self.shuffling {
            self.order.clone()
        } else {
            Vec::new()
        }
    }

    /// Test-only seam: install a presentation order.
    pub fn set_presentation_order_for_test(&mut self, next: Vec<usize>) {
        if self.shuffling {
            self.order = next;
        }
    }

    /// Plays a sequence starting at `start_index` (default 0).
    pub fn play_sequence(
        &mut self,
        items: Vec<PlaybackItem>,
        start_index: i64,
    ) -> Result<PlaybackItem, String> {
        if items.is_empty() {
            return Err("This release has no playable tracks.".to_string());
        }
        let index = start_index.clamp(0, items.len() as i64 - 1);
        self.items = items;
        self.index = index;
        self.history.clear();
        self.rebuild_order();
        self.items
            .get(index as usize)
            .cloned()
            .ok_or_else(|| "This release has no playable tracks.".to_string())
    }

    /// Replaces the queue with a single item.
    pub fn play_now(&mut self, item: PlaybackItem) -> PlaybackItem {
        self.items = vec![item.clone()];
        self.index = 0;
        self.history.clear();
        self.rebuild_order();
        item
    }

    /// Inserts right after the current item.
    pub fn play_next(&mut self, item: PlaybackItem) {
        if self.items.is_empty() {
            self.items = vec![item];
            self.index = 0;
        } else {
            let at = (self.index + 1).max(0) as usize;
            let at = at.min(self.items.len());
            self.items.insert(at, item);
        }
        self.rebuild_order();
    }

    /// Appends to the end of the queue.
    pub fn enqueue(&mut self, item: PlaybackItem) {
        self.items.push(item);
        self.rebuild_order();
    }

    /// Appends several items to the end of the queue.
    pub fn enqueue_many(&mut self, items: Vec<PlaybackItem>) {
        self.items.extend(items);
        self.rebuild_order();
    }

    /// Advances under the active policy. Records history.
    #[allow(clippy::should_implement_trait)]
    pub fn next(&mut self) -> Option<PlaybackItem> {
        let target = self.peek_next_index();
        let target = target?;
        self.index = target as i64;
        self.items.get(target).cloned()
    }

    /// History-aware back navigation.
    pub fn previous(&mut self) -> Option<PlaybackItem> {
        let target = self.peek_prev_index()?;
        self.index = target as i64;
        self.items.get(target).cloned()
    }

    /// Moves the cursor to an item (no queue rebuild).
    pub fn move_to(&mut self, i: i64) {
        if i < 0 || i >= self.items.len() as i64 {
            return;
        }
        self.index = i;
    }

    /// Reorders the canonical list. The cursor follows the current ITEM and
    /// history indices are remapped so Previous still retraces navigation.
    pub fn move_item(&mut self, from: i64, to: i64) {
        let n = self.items.len() as i64;
        if from < 0 || from >= n || to < 0 || to >= n || from == to {
            return;
        }
        let moved = self.items.remove(from as usize);
        self.items.insert(to as usize, moved);
        let remap = |i: i64| -> i64 {
            if i == from {
                return to;
            }
            if from < to && i > from && i <= to {
                return i - 1;
            }
            if from > to && i >= to && i < from {
                return i + 1;
            }
            i
        };
        self.history = self.history.iter().map(|&h| remap(h)).collect();
        self.index = remap(self.index);
        self.rebuild_order();
    }

    /// Removes the item at `i`, adjusting the cursor and history.
    pub fn remove_at(&mut self, i: i64) {
        if i < 0 || i >= self.items.len() as i64 {
            return;
        }
        let removed_cursor_shift = if i < self.index { 1 } else { 0 };
        let was_current = i == self.index;
        self.items.remove(i as usize);
        let mut index = self.index - removed_cursor_shift;
        if was_current {
            index = index.min(self.items.len() as i64 - 1);
        }
        self.history.retain(|&h| h != i);
        self.history = self
            .history
            .iter()
            .map(|&h| if h > i { h - 1 } else { h })
            .collect();
        self.index = index;
        self.rebuild_order();
    }

    /// Empties the queue.
    pub fn clear(&mut self) {
        self.items.clear();
        self.index = -1;
        self.history.clear();
        self.order.clear();
    }

    // -- internal helpers --------------------------------------------------

    fn rebuild_order(&mut self) {
        if !self.shuffling || self.items.is_empty() {
            self.order.clear();
            return;
        }
        let mut upcoming: Vec<usize> = Vec::new();
        for i in 0..self.items.len() {
            if i as i64 != self.index {
                upcoming.push(i);
            }
        }
        let rest = shuffle_order(&upcoming, &mut self.rng);
        self.order = if self.index >= 0 {
            let mut order = Vec::with_capacity(rest.len() + 1);
            order.push(self.index as usize);
            order.extend(rest);
            order
        } else {
            rest
        };
    }

    fn reshuffle(&mut self) {
        self.rebuild_order();
    }

    fn push_history(&mut self, from: i64) {
        if from >= 0 {
            self.history.push(from);
        }
        if self.history.len() > 500 {
            let excess = self.history.len() - 500;
            self.history.drain(0..excess);
        }
    }

    fn peek_next_index(&mut self) -> Option<usize> {
        if self.items.is_empty() {
            return None;
        }
        let cursor = self.index;
        self.push_history(cursor);
        if !self.shuffling {
            return next_index_under_repeat(cursor, self.items.len(), self.repeat);
        }
        let pos = self.order.iter().position(|&x| x as i64 == cursor);
        if let Some(pos) = pos {
            if pos + 1 < self.order.len() {
                return Some(self.order[pos + 1]);
            }
        }
        if self.repeat == RepeatMode::All {
            self.reshuffle();
            return self.order.first().copied();
        }
        None
    }

    fn peek_prev_index(&mut self) -> Option<usize> {
        if self.items.is_empty() {
            return None;
        }
        while let Some(prev) = self.history.pop() {
            if prev >= 0 && prev < self.items.len() as i64 && prev != self.index {
                return Some(prev as usize);
            }
        }
        if !self.shuffling {
            let pi = self.index - 1;
            return if pi >= 0 { Some(pi as usize) } else { None };
        }
        let pos = self.order.iter().position(|&x| x as i64 == self.index);
        pos.and_then(|p| if p > 0 { Some(self.order[p - 1]) } else { None })
    }
}

File: web/player-core/src/queue.ts
// Copyright (c) 2026, The MusicPack Development Team
// SPDX-License-Identifier: BSD-3-Clause

// QueueModel (player-core M3, extended M5).
//
// Framework-free queue with canonical items + cursor. Repeat modes and
// shuffle are POLICY layered over the canonical list (never destroying it):
// - repeat 'one' is resolved by the Player at EOS (reload current), so
//   advance() treats it like 'off';
// - repeat 'all' wraps advance() to index 0;
// - shuffle produces a presentation order via order.ts (seeded RNG);
//   stepBack() consults a history stack so Previous works under shuffle;
//   toggling shuffle off restores canonical order.

import { writable, type Writable } from './store';
import type { PlaybackItem } from './types';
import { nextIndexUnderRepeat, shuffleOrder, type RepeatMode, type Rng } from './order';

export interface QueueState<TItem> {
  items: TItem[];
  /** Index of the current item (-1 when the queue is empty). */
  index: number;
}

function empty<TItem>(): QueueState<TItem> {
  return { items: [], index: -1 };
}

/** Framework-free queue model. Generic over the item shape: hosts may carry
 *  richer entries (the web QueueItem extends PlaybackItem). All semantics
 *  live here; hosts only add item CONSTRUCTION. */
export function createQueueModel<TItem extends PlaybackItem>(opts: { rng: Rng }) {
  const store: Writable<QueueState<TItem>> = writable<QueueState<TItem>>(empty<TItem>());
  /** Presentation order when shuffled: indices into canonical items.
   *  Empty = identity (canonical order). */
  let order: number[] = [];
  /** Visited cursors for history-aware step-back under shuffle/repeat-all. */
  let history: number[] = [];
  const rng: Rng = opts.rng;
  let repeat: RepeatMode = 'off';
  let shuffling = false;

  function effectiveNext(): TItem | null {
    const s = store.get();
    if (s.items.length === 0) return null;
    if (!shuffling) {
      const ni = nextIndexUnderRepeat(s.index, s.items.length, repeat);
      return ni === null ? null : s.items[ni] ?? null;
    }
    // Shuffled: follow the presentation order from the current position.
    const posInOrder = order.indexOf(s.index);
    if (posInOrder >= 0 && posInOrder + 1 < order.length) {
      const ni = order[posInOrder + 1]!;
      return s.items[ni] ?? null;
    }
    // End of the shuffled pass: wrap only under repeat-all (reshuffle).
    if (repeat === 'all') {
      reshuffle();
      const first = order[0];
      return first === undefined ? null : s.items[first] ?? null;
    }
    return null;
  }

  function effectivePrev(): TItem | null {
    const s = store.get();
    if (s.items.length === 0) return null;
    // History-aware back navigation (works under shuffle and after wraps).
    while (history.length > 0) {
      const prev = history.pop()!;
      if (prev >= 0 && prev < s.items.length && prev !== s.index) {
        return s.items[prev] ?? null;
      }
    }
    if (!shuffling) {
      const pi = s.index - 1;
      return pi >= 0 ? s.items[pi] ?? null : null;
    }
    const posInOrder = order.indexOf(s.index);
    if (posInOrder > 0) {
      const pi = order[posInOrder - 1]!;
      return s.items[pi] ?? null;
    }
    return null;
  }

  function rebuildOrder(): void {
    const s = store.get();
    if (!shuffling || s.items.length === 0) {
      order = [];
      return;
    }
    // Current stays first; the rest Fisher–Yates.
    const upcoming: number[] = [];
    for (let i = 0; i < s.items.length; i++) if (i !== s.index) upcoming.push(i);
    const rest = shuffleOrder(upcoming, rng);
    order = s.index >= 0 ? [s.index, ...rest] : [...rest];
  }

  function reshuffle(): void {
    rebuildOrder();
  }

  function pushHistory(from: number): void {
    if (from >= 0) history.push(from);
    // Bound the stack; deep sessions don't need unbounded memory.
    if (history.length > 500) history.splice(0, history.length - 500);
  }

  return {
    ...store,
    state: store.get,

    get repeat(): RepeatMode {
      return repeat;
    },
    setRepeat(mode: RepeatMode): void {
      repeat = mode;
    },

    get shuffle(): boolean {
      return shuffling;
    },
    setShuffle(on: boolean): void {
      if (shuffling === on) return;
      shuffling = on;
      history = [];
      rebuildOrder();
    },
    /** Test/inspection seam: the active presentation order (indices). */
    getPresentationOrder(): number[] {
      return shuffling ? [...order] : [];
    },
    /** Test-only seam: install a presentation order. Lets deterministic
     *  tests pin a shuffle successor without reaching into closures.
     *  Must only be called while shuffle is ON and with a permutation of
     *  canonical indices starting at the cursor. */
    setPresentationOrderForTest(next: number[]): void {
      if (!shuffling) return;
      order = [...next];
    },

    /** Play the given sequence starting at `startIndex` (default first). */
    playSequence(items: TItem[], startIndex = 0): TItem {
      if (items.length === 0) throw new Error('This release has no playable tracks.');
      const index = Math.max(0, Math.min(startIndex, items.length - 1));
      store.set({ items, index });
      history = [];
      rebuildOrder();
      const item = items[index] ?? items[0];
      if (!item) throw new Error('This release has no playable tracks.');
      return item;
    },
    /** Replace the queue with a single item. */
    playNow(item: TItem): TItem {
      store.set({ items: [item], index: 0 });
      history = [];
      rebuildOrder();
      return item;
    },
    /** Insert right after the current item. */
    playNext(item: TItem): void {
      store.update((s) => {
        if (s.items.length === 0) return { items: [item], index: 0 };
        const at = s.index + 1;
        const items = [...s.items];
        items.splice(at, 0, item);
        return { items, index: s.index };
      });
      rebuildOrder();
    },
    /** Append to the end of the queue. */
    enqueue(item: TItem): void {
      store.update((s) => ({ items: [...s.items, item], index: s.index }));
      rebuildOrder();
    },
    enqueueMany(items: TItem[]): void {
      store.update((s) => ({ items: [...s.items, ...items], index: s.index }));
      rebuildOrder();
    },
    current(): TItem | null {
      const s = store.get();
      return s.index >= 0 && s.index < s.items.length ? s.items[s.index] ?? null : null;
    },
    at(i: number): TItem | null {
      return store.get().items[i] ?? null;
    },
    /** Advance under the active policy (repeat/shuffle). Records history. */
    next(): TItem | null {
      let item: TItem | null = null;
      store.update((s) => {
        const target = this.peekNextIndex(s);
        if (target === null) return s;
        const next = { ...s, index: target };
        item = next.items[next.index] ?? null;
        return next;
      });
      return item;
    },
    /** History-aware back navigation. */
    previous(): TItem | null {
      let item: TItem | null = null;
      store.update((s) => {
        const target = this.peekPrevIndex(s);
        if (target === null) return s;
        const next = { ...s, index: target };
        item = next.items[next.index] ?? null;
        return next;
      });
      return item;
    },
    /** Move the cursor to an item (no queue rebuild; user-initiated jump). */
    moveTo(i: number): void {
      store.update((s) => {
        if (i < 0 || i >= s.items.length) return s;
        return { ...s, index: i };
      });
    },
    /** Reorders the canonical list (queue-panel drag/arrows). The cursor
     *  follows the current ITEM across the move; history indices are
     *  remapped so Previous still retraces real navigation. */
    move(from: number, to: number): void {
      store.update((s) => {
        const n = s.items.length;
        if (from < 0 || from >= n || to < 0 || to >= n || from === to) return s;
        const items = [...s.items];
        const [moved] = items.splice(from, 1);
        items.splice(to, 0, moved!);
        const remap = (i: number): number => {
          if (i === from) return to;
          if (from < to && i > from && i <= to) return i - 1;
          if (from > to && i >= to && i < from) return i + 1;
          return i;
        };
        history = history.map(remap);
        return { items, index: remap(s.index) };
      });
      rebuildOrder();
    },
    removeAt(i: number): void {
      store.update((s) => {
        if (i < 0 || i >= s.items.length) return s;
        const removedCursorShift = i < s.index ? 1 : 0;
        const wasCurrent = i === s.index;
        const items = s.items.filter((_, idx) => idx !== i);
        let index = s.index - removedCursorShift;
        if (wasCurrent) index = Math.min(index, items.length - 1);
        // keep history pointing at valid entries
        history = history.filter((h) => h !== i).map((h) => (h > i ? h - 1 : h));
        return { items, index };
      });
      rebuildOrder();
    },
    clear(): void {
      store.set(empty<TItem>());
      history = [];
      order = [];
    },

    // -- internal helpers used by next()/previous() above ------------------
    peekNextIndex(s: QueueState<TItem>): number | null {
      if (s.items.length === 0) return null;
      pushHistory(s.index);
      if (!shuffling) return nextIndexUnderRepeat(s.index, s.items.length, repeat);
      const posInOrder = order.indexOf(s.index);
      if (posInOrder >= 0 && posInOrder + 1 < order.length) return order[posInOrder + 1]!;
      if (repeat === 'all') {
        reshuffle();
        return order[0] ?? null;
      }
      return null;
    },
    peekPrevIndex(s: QueueState<TItem>): number | null {
      if (s.items.length === 0) return null;
      while (history.length > 0) {
        const prev = history.pop()!;
        if (prev >= 0 && prev < s.items.length && prev !== s.index) return prev;
      }
      if (!shuffling) {
        const pi = s.index - 1;
        return pi >= 0 ? pi : null;
      }
      const posInOrder = order.indexOf(s.index);
      return posInOrder > 0 ? order[posInOrder - 1]! : null;
    },
  };
}

export type QueueModel<TItem extends PlaybackItem = PlaybackItem> = ReturnType<
  typeof createQueueModel<TItem>
>;
